Ledger tokens with memo

fmorency opened the pull request:
There's still some `TODO` and code cleanup but I think we can start the review process

This PR contains
- The `tokens.create` endpoint
- The `tokens.update` endpoint
- The `tokens.info` endpoint
- The `tokens.addExtendedInfo` endoint
- The `tokens.removeExtendedInfo` endpoint
- Integration tests (cucumber) for the `tokens.*` endpoints
- E2E tests (bats) for the `tokens.*` endpoints
- A refactor of the `account_id` subresource counter to a more general `subresource_id` counter
- A Token Migration
  - A token metadata to the storage
  - Port the `accound_id` subresource counter to `subresource_id`
- Resiliency tests (bats) for the `tokens.*` endpoint
- Resiliency tests (bats) for the Token Migration
- A `tokens` subcommand for the `ledger` CLI
- Optional symbol metadata in the staging file
- Ledger `send()` now support a memo (TODO: add a migration for that?)
- An alternative method to fetch the token symbols supported by the server
  - Only available when Token Migration is active
  - No CBOR decoding involved
- A symbol storage iterator
  - Only useful when Token Migration is active 
- Ability to skip the staging file hash check in tests
- Some token helper functions for BATs tests

Depends on https://github.com/liftedinit/many-rs/pull/223

Fixes liftedinit/roadmap#21
Fixes liftedinit/roadmap#62

hansl committed:
Refactor the Memo field into a legacy memo and ng memo

This follows the many-rs PR https://github.com/liftedinit/many-rs/pull/201
and the specification latest master and its follow up PR https://github.com/many-protocol/specification/pull/58
which includes burnt fields (to add new fields instead of re-typing
old ones).

hansl committed:
.

hansl committed:
Update migrations to latest version of many-rs migration

hansl committed:
Clippy

hansl committed:
Merge branch 'main' into memo-refactor

hansl committed:
.

hansl committed:
.

hansl committed:
.

hansl committed:
.

hansl committed:
.

hansl committed:
.

hansl committed:
.

hansl committed:
.

hansl committed:
.

hansl committed:
Add a resiliency test

fmorency committed:
Fix grep stray backslash warning

fmorency committed:
Merge remote-tracking branch 'origin/main' into memo-refactor

fmorency committed:
Fix memo migration test

fmorency committed:
Add a bunch of missing Memo for Send

fmorency committed:
Implement most of the token creation features

Still missing real symbol generation, ACL and more tests

fmorency committed:
Fix test staging files

fmorency committed:
Initial support for token creation in ledger binary

fmorency committed:
Add ledger token module

fmorency committed:
.

fmorency committed:
.

fmorency committed:
More testing and ACL

fmorency committed:
Test cleanup

fmorency committed:
More tests

fmorency committed:
More tests

fmorency committed:
More tests

fmorency committed:
Clippy

fmorency committed:
Merge remote-tracking branch 'origin/main' into ledger-tokens-with-memo

fmorency committed:
Disable failing test, for now

fmorency committed:
Enhance ledger CLI

fmorency committed:
DRY

fmorency committed:
.

fmorency committed:
Update Bazel with new files

fmorency committed:
Better ext info handling in CLI

fmorency committed:
Empty ext. info for MFX

fmorency committed:
Fix token update permission

fmorency committed:
Formatting

fmorency committed:
Fix clippy and revert use of eprintln

fmorency committed:
Merge branch 'fmorency/main-fix-clippy' into ledger-tokens-with-memo

fmorency committed:
.

fmorency committed:
Fix symbol update

- Create new symbol storage iterator
- Move DB iterator to its own file

fmorency committed:
Cleanup formatting

fmorency committed:
Fix memo in TokenUpdate test

fmorency committed:
.

fmorency committed:
Tokens e2e tests

fmorency committed:
Finish token e2e tests and code refactor

fmorency committed:
Fix tests

fmorency committed:
Fix typo

fmorency committed:
Token migration draft

fmorency committed:
Cleanup token migration and resiliency tests

fmorency commented:
@hansl still some code cleanup to do, but I think you can start reviewing this.

fmorency committed:
Refactor shared test code

fmorency committed:
.

fmorency committed:
.

fmorency committed:
DRY with macro

fmorency committed:
many-ledger test-macros in Bazel

fmorency committed:
.

fmorency committed:
Fix many TODO

Custom error and cleanup

fmorency committed:
.

hansl commented on src/many-ledger/src/json.rs:
```diff
@@ -126,12 +126,21 @@ impl AccountJson {
 }
 
+#[derive(serde::Deserialize, Clone, Debug)]
+pub struct SymbolMetaJson {
```
Add memo please. It _should_ serialize using Serde (and if not, not a bad idea to add it).

fmorency replied:
A `memo` for what exactly?

The `tokens.create` endpoint doesn't have a `memo` field.

fmorency replied:
The `memo` for the "other calls" is logged in the event. The only token creation endpoint with this behavior is `tokens.update`.

The `memo` in `token-extended-info` is part of the extended info and can be retrieved using `tokens.info`. It is NOT part of the event. Only the extended info attribute indices are.

How would you like to handle this?

hansl commented on src/many-ledger/src/migration/tokens.rs:
```diff
@@ -0,0 +1,162 @@
+        "symbol",
+        "symbol_name",
+        "symbol_decimals",
+        "symbol_total",       // TODO: This field could be calculated from the storage
```
TODO unnecessary. Denormalizing your data in KVStore patterns is not unusual. If you want to keep a comment explaining why, go ahead. KVStore don't have the join query speed of RDBMs.

hansl commented on src/many-ledger/src/module/ledger_tokens.rs:
```diff
@@ -0,0 +1,197 @@
+        args: TokenCreateArgs,
+    ) -> Result<TokenCreateReturns, ManyError> {
+        if !self.storage.migrations().is_active(&TOKEN_MIGRATION) {
+            return Err(ManyError::unknown(
```
You should return `InvalidMethodName` instead, which is more expected if a method isn't implemented.

hansl commented on src/many-ledger/src/module/ledger_tokens.rs:
```diff
@@ -0,0 +1,197 @@
+        sender: &Address,
+        args: TokenUpdateArgs,
+    ) -> Result<TokenUpdateReturns, ManyError> {
+        // TODO: Limit token update to given sender
```
Is this going to be a problem? Updates are already limited to owners.

hansl commented on src/many-ledger/src/module/ledger_tokens.rs:
```diff
@@ -0,0 +1,197 @@
+        sender: &Address,
+        args: TokenAddExtendedInfoArgs,
+    ) -> Result<TokenAddExtendedInfoReturns, ManyError> {
+        // TODO: Limit adding extended info to given sender
```
Same as above. Maybe only creation should be limited.

fmorency replied:
Limit creation to MFX address

hansl commented on src/many-ledger/src/storage/ledger_tokens.rs:
```diff
@@ -0,0 +1,336 @@
+        } = args;
+
+        // Create a new token symbol and store in memory and in the persistent store
+        let symbol = self.new_subresource_id();
```
Let's chat about this offline a bit.

fmorency replied:
Use MFX address subresources for this.

fmorency committed:
Merge remote-tracking branch 'origin/main' into ledger-tokens-with-memo

fmorency committed:
Create `TokenMaybeOwner` from CLI `str`

fmorency committed:
Remove TODO

fmorency committed:
Update error type

fmorency committed:
Fix from review

fmorency committed:
Merge remote-tracking branch 'origin/main' into ledger-tokens-with-memo

fmorency committed:
clippies

fmorency committed:
Re-add Dummy migration config for CI

fmorency committed:
Fix TODO

fmorency committed:
Fix CI, maybe

Fix Bazel tests

fmorency committed:
Refactoring

fmorency committed:
Remove dev stuff

fmorency committed:
Remove more `unwrap` and `expect`

fmorency committed:
Use `wait_for_block` instead of `sleep`

fmorency committed:
.

fmorency committed:
Merge remote-tracking branch 'origin/main' into ledger-tokens-with-memo

fmorency committed:
More `map_err` refactoring

fmorency committed:
Better errors

fmorency committed:
More refactor

hansl reviewed (changes requested):
Still going through it, but first round of comments.

fmorency committed:
Merge remote-tracking branch 'origin/main' into ledger-tokens-with-memo

fmorency committed:
Some fix from review

- Use mime_guess for image logo content type identification
- Add token_identity to docker ledger state
- Re-add hash to docker ledger state

fmorency commented on src/many-kvstore/src/storage.rs:
```diff
@@ -38,8 +38,8 @@ pub struct KvStoreStorage {
-    next_account_id: u32,
-    account_identity: Address,
+    next_subresource: u32,
+    root_identity: Address,
```
I could add `account_identity` as a staging parameter, in the same fashion as `token_identity`, and default to the server `identity` if not present.

WDYT?

fmorency commented on src/many-kvstore/src/storage.rs:
```diff
@@ -58,19 +58,32 @@ impl KvStoreStorage {
-        let next_account_id = persistent_store
-            .get(b"/config/account_id")
+        let next_subresource = persistent_store
+            .get(b"/config/subresource_id")
```
See https://github.com/liftedinit/many-framework/pull/293#discussion_r1069787719

fmorency committed:
Refactor initial_distribution in ledger

fmorency commented on src/ledger/src/tokens.rs:
```diff
@@ -0,0 +1,294 @@
+    #[clap(value_parser = token_maybe_owner)]
+    owner: Option<TokenMaybeOwner>,
+
+    // TODO: How to make this better?
```
I pushed a refactor :)

fmorency committed:
.

fmorency committed:
.

fmorency committed:
Add token info balance test

fmorency committed:
Fix token info summary crash

fmorency committed:
Check balance in create new token e2e test

fmorency committed:
Fix token identity propagation after migration

fmorency committed:
.

hansl reviewed (changes requested):
Specification was also updated to add a memo field.

hansl commented on src/many-ledger/src/storage/ledger_tokens.rs:
```diff
@@ -0,0 +1,522 @@
+        };
+
+        // Create the token information and store it in the persistent storage
+        let maybe_owner = owner
+            .as_ref()
+            .map_or(Some(*sender), |maybe_owner| match maybe_owner {
+                Either::Left(addr) => Some(*addr),
+                Either::Right(_) => None,
+            });
```
```suggestion
        let maybe_owner = owner.map_or(Some(*sender), |maybe_owner| maybe_owner.left());
```

fmorency committed:
Add memo to token create|(add|remove)ExtendedInfo

- Add tests covering this support

fmorency committed:
Re-enable staging file hash

fmorency committed:
Update src/many-ledger/test-utils/Cargo.toml

Co-authored-by: Hans Larsen <hans@larsen.online>
Signed-off-by: Félix C. Morency <1102868+fmorency@users.noreply.github.com>

fmorency committed:
Update src/many-ledger/test-utils/Cargo.toml

Co-authored-by: Hans Larsen <hans@larsen.online>
Signed-off-by: Félix C. Morency <1102868+fmorency@users.noreply.github.com>

fmorency committed:
Update src/many-ledger/test-macros/Cargo.toml

Co-authored-by: Hans Larsen <hans@larsen.online>
Signed-off-by: Félix C. Morency <1102868+fmorency@users.noreply.github.com>

fmorency committed:
Update src/many-ledger/test-macros/Cargo.toml

Co-authored-by: Hans Larsen <hans@larsen.online>
Signed-off-by: Félix C. Morency <1102868+fmorency@users.noreply.github.com>

fmorency committed:
Update src/many-ledger/test-macros/Cargo.toml

Co-authored-by: Hans Larsen <hans@larsen.online>
Signed-off-by: Félix C. Morency <1102868+fmorency@users.noreply.github.com>

fmorency committed:
Fix typo in cucumber test tag

fmorency committed:
Merge branch 'ledger-tokens-with-memo' of github.com:fmorency/many-framework into ledger-tokens-with-memo

fmorency committed:
Update crate name after change

fmorency commented on src/many-ledger/src/storage/ledger_tokens.rs:
```diff
@@ -0,0 +1,522 @@
+
+    /// Fetch symbols from `/config/symbols/{symbol}`
+    ///     No CBOR decoding needed.
+    pub(crate) fn _get_symbols(&self, symbols: &mut BTreeSet<Symbol>) -> Result<(), ManyError> {
```
You're right, good catch.

fmorency committed:
Update src/many-ledger/src/storage/ledger_tokens.rs

Co-authored-by: Hans Larsen <hans@larsen.online>
Signed-off-by: Félix C. Morency <1102868+fmorency@users.noreply.github.com>

fmorency committed:
Update src/many-ledger/src/storage/ledger_tokens.rs

Co-authored-by: Hans Larsen <hans@larsen.online>
Signed-off-by: Félix C. Morency <1102868+fmorency@users.noreply.github.com>

hansl commented on src/many-ledger/src/storage/ledger_tokens.rs:
```diff
@@ -0,0 +1,519 @@
+    format!("/config/ext_info/{symbol}").into_bytes()
+}
+
+pub struct SymbolMeta {
```
This is the same as `SymbolMetaJson`. Any reason not to use the same type?

fmorency replied:
I want to separate the configuration file layout from our internal structure layout. It is the same right now, but it might not be the case soon™.

hansl commented on src/many-ledger/src/storage.rs:
```diff
@@ -90,41 +104,66 @@ impl LedgerStorage {
         &self.migrations
     }
 
+    fn subresource_db_key(migration_is_active: bool) -> Vec<u8> {
```
Can you make this a function local to the file but outside the LedgerStorage type?

fmorency replied:
This is getting nuked with the `account_identity` refactor.

fmorency committed:
Update src/many-ledger/tests/features/ledger_tokens/remove_token_ext_info.feature

Co-authored-by: Hans Larsen <hans@larsen.online>
Signed-off-by: Félix C. Morency <1102868+fmorency@users.noreply.github.com>

hansl commented on src/many-ledger/src/migration/tokens.rs:
```diff
@@ -0,0 +1,190 @@
+        .ok_or_else(|| ManyError::unknown(format!("Symbol {symbol} not found in DB")))
+        .cloned()?;
+
+    let owner: Option<Address> = serde_json::from_value(extra["symbol_owner"].clone())
+        .map_err(ManyError::deserialization_error)?;
+    let maximum: Option<TokenAmount> = serde_json::from_value(extra["symbol_maximum"].clone())
+        .map_err(ManyError::deserialization_error)?;
+    let name: String = serde_json::from_value(extra["symbol_name"].clone())
+        .map_err(ManyError::deserialization_error)?;
+    let decimals: u64 = serde_json::from_value(extra["symbol_decimals"].clone())
+        .map_err(ManyError::deserialization_error)?;
+    let total: TokenAmount = serde_json::from_value(extra["symbol_total"].clone())
+        .map_err(ManyError::deserialization_error)?;
+    let circulating: TokenAmount = serde_json::from_value(extra["symbol_circulating"].clone())
+        .map_err(ManyError::deserialization_error)?;
+    let info = TokenInfo {
+        symbol,
+        summary: TokenInfoSummary {
+            name,
+            ticker,
+            decimals,
+        },
+        supply: TokenInfoSupply {
+            total,
+            circulating,
+            maximum,
+        },
+        owner,
+    };
```
I tried to make it easier to follow and less repetitive. Best would really be to have a type that directly deserialize, but 🤷 

```suggestion
    let info = (move || {
        Ok::<_, serde_json::Error>(TokenInfo {
            symbol,
            summary: TokenInfoSummary {
                name: serde_json::from_value(extra["symbol_name"].clone())?,
                ticker,
                decimals: serde_json::from_value(extra["symbol_decimals"].clone())?,
            },
            supply: TokenInfoSupply {
                total: serde_json::from_value(extra["symbol_total"].clone())?,
                circulating: serde_json::from_value(extra["symbol_circulating"].clone())?,
                maximum: serde_json::from_value(extra["symbol_maximum"].clone())?,
            },
            owner: serde_json::from_value(extra["symbol_owner"].clone())?,
        })
    })()
    .map_err(ManyError::deserialization_error)?;
```

fmorency replied:
That's cute!

hansl commented on src/many-ledger/src/migration/tokens.rs:
```diff
@@ -0,0 +1,190 @@
+    };
+
+    // Add token data to the DB
+    storage
```
Is there a reason this isn't a single apply for all the keys?

hansl commented:
Another comment I'd like to see is coverage given by cucumber. It seems to be missing.

fmorency committed:
Refactor `account_identity`

fmorency committed:
Merge branch 'ledger-tokens-with-memo' of github.com:fmorency/many-framework into ledger-tokens-with-memo

fmorency committed:
Update src/many-ledger/src/migration/tokens.rs

Co-authored-by: Hans Larsen <hans@larsen.online>
Signed-off-by: Félix C. Morency <1102868+fmorency@users.noreply.github.com>

fmorency committed:
Merge branch 'ledger-tokens-with-memo' of github.com:fmorency/many-framework into ledger-tokens-with-memo

fmorency committed:
Fix potential side effect in `get_symbols`

fmorency committed:
Single apply for all keys

fmorency commented on src/many-ledger/src/storage/account.rs:
```diff
@@ -6,37 +7,67 @@ use crate::storage::multisig::{
-        self.account_identity
-            .with_subresource_id(current_id)
-            .expect("Too many accounts")
+                if account.subresource_id.is_some()
```
I think it'll crash in the `Err`.

Using `id.subresource_id().is_some() && account.subresource_id != id.subresource_id()` as the condition, we enter the if if `id.subresource_id() == Some(_)` and `account.subresource_id == None`, which leads to a crash in

```
          return Err(error::unexpected_subresource_id(
                        id.subresource_id().unwrap().to_string(),
                        account.subresource_id.unwrap().to_string(),
                    ));
```

caused by `account.subresource_id.unwrap().to_string()`

fmorency commented on tests/resiliency/ledger/token_migration.bats:
```diff
@@ -0,0 +1,172 @@
+    mkdir "$BATS_TEST_ROOTDIR"
+
+    echo '
+    { "migrations": [
```
https://github.com/liftedinit/many-framework/issues/320

fmorency committed:
Clippy

fmorency commented:
You can check the coverage at https://app.codecov.io/gh/liftedinit/many-framework/pull/293

Here's some screenshots from our `make code-coverage` command

![2023-01-20_11-12](https://user-images.githubusercontent.com/1102868/213748547-de9ce34b-836e-4dc9-9795-a8619bb9a2da.png)
![2023-01-20_11-13](https://user-images.githubusercontent.com/1102868/213748579-7d999f43-10a1-42e7-8cee-a44fb430be09.png)
![2023-01-20_11-14](https://user-images.githubusercontent.com/1102868/213748769-6868841e-53f3-456b-a485-ef2bd059b2d5.png)
![2023-01-20_11-13_1](https://user-images.githubusercontent.com/1102868/213748662-8ea49075-361e-4e84-bb93-d0c0099290db.png)

hansl reviewed (approved):
LGTM.

fmorency committed:
Update to latest `many-rs`

fmorency committed:
Merge remote-tracking branch 'origin/main' into ledger-tokens-with-memo

fmorency committed:
Fix merk
